### PR TITLE
@TransactionScoped Context does not call @Predestroy on TransactionScoped Beans

### PR DESCRIPTION
**Pull request by msfm**

### CODE REVIEW

**Comment by vhmolinar** [comment]
This fix is very important for me.
Any chances to have it in the next release? @msfm

**Comment by msfm** [comment]
@vhmolinar This needs to wait for a review and approval.

File: integration-tests/narayana-jta/src/main/java/io/quarkus/narayana/jta/TransactionScopedBean.java
Comment (Sanne) [inline integration-tests/narayana-jta/src/main/java/io/quarkus/narayana/jta/TransactionScopedBean.java:13]: these two fields need to be `final`

File: integration-tests/narayana-jta/src/main/java/io/quarkus/narayana/jta/TransactionScopedBean.java
Comment (Sanne) [inline integration-tests/narayana-jta/src/main/java/io/quarkus/narayana/jta/TransactionScopedBean.java:10]: Could you give this a different name? You're adding two classes with this name; I understand it's not a problem as they are in a different package, but it's a bit unpractical to have two around.

File: extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/context/TransactionContext.java
Comment (Sanne) [inline extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/context/TransactionContext.java:36]: I'd prefer this to use a `ConcurrentMap`, so to not need `synchronized` on it further below.

File: extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/context/TransactionContext.java
Comment (Sanne) [inline extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/context/TransactionContext.java:173]: after making the `transactions` Map a `ConcurrentMap`, the synchronized can be removed.

File: extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/context/TransactionContext.java
Comment (Sanne) [inline extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/context/TransactionContext.java:117]: after making the `transactions` Map a `ConcurrentMap`, the synchronized can be removed; the following operations can be made atomic with a little bit of care, e.g. using `computeIfAbsent` instead of `get`.
Even better, for optimal performance, to a `get` first, and then resort to `computeIfAbsent` only if the first optimistic get operation failed.

**Review by Sanne — CHANGES_REQUESTED** [review]
Thanks for the fix!

I think it's a good approach but there's room to improve its concurrency; I've made some suggestions.

**Comment by mkouba** [comment]
I can't say I fully understand all the details of the `TransactionContext` impl but I'd say that it should be enough to change the `TransactionContext.TransactionContextState.remove(Contextual<T>)` to something like:
```java
<T> void remove(Contextual<T> bean) {
   ContextInstanceHandle<?> instance = mapBeanToInstanceHandle.remove(bean);
   if (instance != null) {
      instance.destroy();
   }
}
```
AND ensure the `io.quarkus.narayana.jta.runtime.context.TransactionContext.destroy()` is called when the transaction is over (commited/rolled back).

CC @ochaloup @manovotn

**Comment by ochaloup** [comment]
I don't remember all the details but the issue of missing the functionality seems to be caused by fact of not porting the `destroy` functionality from the Narayana CDI extension.
In the current shape of the code I agree with @mkouba at https://github.com/quarkusio/quarkus/pull/14053#issuecomment-754463844 that the call should be enough.

@msfm would be possible to check if the change that @mkouba suggests is fine for the `@PreDestroy` test passing?

**Comment by mkouba** [comment]
> I don't remember all the details but the issue of missing the functionality seems to be caused by fact of not porting the `destroy` functionality from the Narayana CDI extension.
> In the current shape of the code I agree with @mkouba at [#14053 (comment)](https://github.com/quarkusio/quarkus/pull/14053#issuecomment-754463844) that the call should be enough.
> 
> @msfm would be possible to check if the change that @mkouba suggests is fine for the `@PreDestroy` test passing?

The problem is that `TransactionContext.destroy()` is probably not called when the TX is over. We'll probably need to register a JTA Synchronization or something like that?

**Comment by Sanne** [comment]
> The problem is that `TransactionContext.destroy()` is probably not called when the TX is over. We'll probably need to register a JTA Synchronization or something like that?

It was my understand that that's exactly what this PR does?

**Comment by mkouba** [comment]
> > The problem is that `TransactionContext.destroy()` is probably not called when the TX is over. We'll probably need to register a JTA Synchronization or something like that?
> 
> It was my understand that that's exactly what this PR does?

It does but it also adds some redundant structures such as `TransactionScopeCleanup`...

**Comment by zhfeng** [comment]
I belived that most of the codes are back porting from the Narayana CDI implementation. So in quarkus, I agree with @manovotn and @mkouba that we need to register a ```Synchronization``` to call the ```destory()```.

```
private static class TransactionContextState implements ContextState, Synchronization {
    public TransactionContextState(Transaction transaction) {
        transaction.registerSynchronization(this);
    }
   
    @Override
    public void beforeCompletion() {
    }

    @Override
    public void afterCompletion(int i) {
        this.destory();
    }
}
```

**Comment by ochaloup** [comment]
@mkouba yes, you're right
@zhfeng the approach seems good to me

**Comment by msfm** [comment]
Thank you all for the review and valuable feedback!

The patch which I proposed was simply porting `com.arjuna.ats.jta.cdi.TransactionScopeCleanup` and `com.arjuna.ats.jta.cdi.TransactionContext` from narayana. (and the original implementation was incorporated into narayana as a fix for [JBTM-1983](https://issues.redhat.com/browse/JBTM-1983).)

I've updated this fix per @zhfeng's suggestion. Also made changes to TransactionScopedBean in the unit test per @Sanne's suggestion. (By the way, I haven't changed the class name TransactionScopedBean yet, since the same class name no longer exists in the new patch. Also, the class name TransactionScopedBean has existed before this patch, so I'd like to leave it as it is.)

File: extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/context/TransactionContext.java
Comment (ochaloup) [inline extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/context/TransactionContext.java:230]: Despite the argument is not used it could be named better.

```suggestion
        public void afterCompletion(int status) {
```

**Review by ochaloup — APPROVED** [review]
To me this seems good (adding just a minor point about naming).

**Review by zhfeng — APPROVED** [review]

File: extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/context/TransactionContext.java
Comment (msfm) [reply extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/context/TransactionContext.java:230]: Got it. I'll incorporate the suggestion.

**Review by mkouba — CHANGES_REQUESTED** [review]
@msfm You still need to modify the `TransactionContext.TransactionContextState.remove()` method because it's used when `AlterableContext#destroy()` method is used to destroy an existing contextual instance.

**Comment by msfm** [comment]
> @msfm You still need to modify the `TransactionContext.TransactionContextState.remove()` method because it's used when `AlterableContext#destroy()` method is used to destroy an existing contextual instance.

Ah, I'm sorry that I overlooked the suggestion you pointed out in the previous comment. I'll update this PR with the following change.

```
--- a/extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/context/TransactionContext.java
+++ b/extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/context/TransactionContext.java
@@ -188,7 +188,10 @@ public class TransactionContext implements InjectableContext {
          * @param bean contextual bean instance
          */
         <T> void remove(Contextual<T> bean) {
-            mapBeanToInstanceHandle.remove(bean);
+            ContextInstanceHandle<?> instance = mapBeanToInstanceHandle.remove(bean);
+            if (instance != null) {
+                instance.destroy();
+            }
         }
```

**Review by mkouba — APPROVED** [review]

**Review by Sanne — APPROVED** [review]
Nice, much simpler and efficient.

**Comment by gsmet** [comment]
@Ladicek ping for unstable MP FT tests. They are failing very often so either we can fix them or we need to disable them for the time being.

**Comment by Ladicek** [comment]
Hmm, you mean the MP FT TCK it seems. It's been a long while since I've seen a TCK test failing in the SmallRye Fault Tolerance repo, so I wonder what's different here. Hmm, it might be because I fixed a bunch of issues in the 5.0 version, which is still only in the `microprofile-4` branch...

Well, if it's failing "very often", as you say, then feel free to disable them for now, and we can reenable with the update to SRye FT 5.0.

**Comment by gsmet** [comment]
Thanks everyone!

**Pull request merged**
